tracepointdebug/probe/error_stack_manager.py

```diff
@@ -0,0 +1,123 @@
+import time, traceback
+from tracepointdebug.probe.coded_exception import CodedException
+from tracepointdebug.probe.coded_exception import CodedException
+from tracepointdebug.probe.event.errorstack.error_stack_rate_limit_event import ErrorStackRateLimitEvent
+from tracepointdebug.probe.event.errorstack.error_stack_snapshot_event import ErrorStackSnapshotEvent
+from tracepointdebug.probe.event.errorstack.error_stack_snapshot_failed_event import ErrorStackSnapshotFailedEvent
+from tracepointdebug.probe.ratelimit.rate_limit_result import RateLimitResult
+from tracepointdebug.probe.ratelimit.rate_limiter import RateLimiter
+from tracepointdebug.probe.snapshot_collector import SnapshotCollector
+import logging, sys, threading
+from tracepointdebug.config import config_names
+from tracepointdebug.config.config_provider import ConfigProvider
+from datetime import datetime as dt
+from cachetools import TTLCache
+import datetime, os
+
+logger = logging.getLogger(__name__)
+
+_MAX_SNAPSHOT_SIZE = 32768
+_MAX_FRAMES = 10
+_MAX_EXPAND_FRAMES = 2
+_MAX_TIME_TO_ALIVE_MIN = 5
+
+class ErrorStackManager(object):
+    __instance = None
+
+    def __init__(self, broker_manager):
+        self.broker_manager = broker_manager
+        self.old_settrace = sys.gettrace()
+        self.old_threading = threading._trace_hook
+        self.condition = None
+        self.timer = None
+        self.sidekick_exception = "sidekickException"
+        self.rate_limiter = RateLimiter()
+        self.ttl_cache = TTLCache(maxsize=2048, ttl=datetime.timedelta(minutes=_MAX_TIME_TO_ALIVE_MIN), timer=datetime.datetime.now)
+        ErrorStackManager.__instance = self
+
+    @staticmethod
+    def instance(*args, **kwargs):
+        return ErrorStackManager(*args,
+                                 **kwargs) if ErrorStackManager.__instance is None else ErrorStackManager.__instance
+
+    @staticmethod
+    def get_id(file, line):
+        return '{}:{}:{}'.format(file, line, str(dt.now()))
+
+    def _get_point_cache_id(self, frame):
+        return frame.f_code.co_filename + ":::" + str(frame.f_lineno)
+
+    def _check_point_inserted(self, frame):
+        error_point_id = self._get_point_cache_id(frame)
+        item = self.ttl_cache.get(error_point_id, None)
+        if item is None:
+            self.ttl_cache[error_point_id] = True
+            return False
+        return True
+
+    def _white_list_exceptions(self, frame):
+        frame_file_path = os.path.abspath(frame.f_code.co_filename)
+        blacklist = ["python", "site-packages", "importlib", "tracepointdebug"]
+        for black in blacklist:
+            if black in frame_file_path:
+                return False
+        return True
+
+    def trace_hook(self, frame, event, arg):
+        if not self._white_list_exceptions(frame):
+            return
+        frame.f_trace = self._frame_hook
+
+    def _frame_hook(self, frame, event, arg):
+        try:
+            if event != "exception":
```

Comment:
It might be better in terms of performance to move this check `f event != "exception":` into  `trace_hook` so for non exception cases, `_frame_hook` will not be called at all.